### PR TITLE
Fix SRL hang when exit.

### LINKED ISSUE #286: train.sh in semantic_role_labeling cannot exit at the end
It hangs when the last pass is trained, for the following settings:
1. use_gpu = false, trainer_count=4
2. use_gpu = true, trainer_count=4 
3. use_gpu = true, trainer_count=1

It can exit normally only when
4. use_gpu = false, trainer_count=1

### PR DESCRIPTION
**Pull request by reyoung**
* Error occurred when enable Async Load in TestDataProvider.
  * It because DataProvider is calling getNextBatchInternal in one thread, and destructing DataProvider in other thread.
  * Add wait routine in DataProvider destructing.
* Also fix another bug, when destructing TestDataProvider and do not read any test data.

Fix #286

### CODE REVIEW

**Comment by reyoung** [comment]
@qingqing01 @hedaoyuan This patch is done. It can be reviewed now.

File: paddle/gserver/dataproviders/PyDataProvider2.cpp [hunk 1]
```diff
@@ -529,6 +536,7 @@ class PyDataProvider2 : public DataProvider {
    * Loading a batch of data.
    */
   int64_t getNextBatchInternal(int64_t size_, DataBatch *batch) {
+    auto guard = this->guard();
```
Comment (hedaoyuan) [inline paddle/gserver/dataproviders/PyDataProvider2.cpp]: It looks like mutex can support it. Why need guard?

File: paddle/gserver/dataproviders/DataProvider.cpp [hunk 2]
```diff
@@ -131,9 +131,10 @@ void DoubleBuffer::asyncLoadBatch() {
     taskReadySem_.wait();
     if (stopping_) break;
 
-    while (batchSize_ == 0) {
+    while (batchSize_ == 0 && !stopping_) {
```
Comment (reyoung) [inline paddle/gserver/dataproviders/DataProvider.cpp]: 有一种情况，即dataprovider一次都没被调用过，就直接退出。。batch size是0，stopping是true

**Comment by reyoung** [comment]
确实，DoubleBuffer不太适合作为DataProvider的member。因为在析构的时候，是子类先析构，基类后析构。所以DoubleBuffer自己的在析构的时候，DataProvider已经析构了一半了。

比如，DoubleBuffer和PyDataProvider组合的情况，内存里有三个对象，DoubleBuffer* PyDataProvider*, DataProvider*。三者析构的顺序是，PyDataProvider*, DoubleBuffer*, DataProvider*。而PyDataProvider*和DataProvider*是一个对象。

我觉得正确的实现DoubleBuffer应该实现一个类似于装饰者模式的东西。这样就不会有这种奇怪的析构顺序。

不过可以先hotfix掉这个bug，然后我们再下一个milestone的时候把这个DoubleBuffer重构一下。

**Comment by pineking** [comment]
刚验证了下，解决了之前训练完成 exit 时， hang 住的问题
测试的是基于 CPU 版

File: paddle/gserver/dataproviders/PyDataProvider2.cpp (hunk 1 above)
Comment (emailweixu) [inline paddle/gserver/dataproviders/PyDataProvider2.cpp]: yes, mutex should work

**Comment by reyoung** [comment]
Related #313

**Comment by reyoung** [comment]
@hedaoyuan I use mutex now, follow comments. It could be reviewed again.

**Pull request merged**